modules/storage/filesystem/btrfs.go

```diff
@@ -0,0 +1,343 @@
+package filesystem
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/threefoldtech/zosv2/modules"
+)
+
+var (
+	_ Filesystem = (*btrfs)(nil)
+
+	DeviceAlreadyMountedError = fmt.Errorf("device is already mounted")
+	DeviceNotMountedError     = fmt.Errorf("device is not mounted")
+)
+
+// btrfs is the filesystem implementation for btrfs
+type btrfs struct {
+	devices DeviceManager
+}
+
+// NewBtrfs creates a new filesystem that implements btrfs
+func NewBtrfs(manager DeviceManager) Filesystem {
+	return &btrfs{manager}
+}
+
+func (b *btrfs) btrfs(ctx context.Context, args ...string) ([]byte, error) {
+	return run(ctx, "btrfs", args...)
+}
+
+func (b *btrfs) Create(ctx context.Context, name string, devices []string, policy modules.RaidProfile) (Pool, error) {
+	name = strings.TrimSpace(name)
+	if len(name) == 0 {
+		return nil, fmt.Errorf("invalid name")
+	}
+
+	block, err := b.devices.WithLabel(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(block) != 0 {
+		return nil, fmt.Errorf("unique name is required")
+	}
+
+	for _, device := range devices {
+		dev, err := b.devices.Device(ctx, device)
+		if err != nil {
+			return nil, err
+		}
+
+		if dev.Used() {
+			return nil, fmt.Errorf("device '%s' is already used", dev)
+		}
+	}
+
+	args := []string{
+		"-L", name,
+		"-d", string(policy),
+		"-m", string(policy),
+	}
+
+	args = append(args, devices...)
+	if _, err := run(ctx, "mkfs.btrfs", args...); err != nil {
+		return nil, err
+	}
+
+	return btrfsPool(name), nil
+}
+
+func (b *btrfs) List(ctx context.Context) ([]Pool, error) {
+	var pools []Pool
+	available, err := BtrfsList(ctx, "", false)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, fs := range available {
+		if len(fs.Label) == 0 {
+			// we only assume labeled devices are managed
+			continue
+		}
+
+		pools = append(pools, btrfsPool(fs.Label))
+	}
+
+	return pools, nil
+}
+
+type btrfsPool string
+
+// Mounted checks is the pool is mounted
+// It doesn't check the default mount location of the pool
+// but instead check if any of the pool devices is mounted
+// under any location
+func (p btrfsPool) Mounted() (string, bool) {
+	ctx := context.Background()
+	list, _ := BtrfsList(ctx, string(p), true)
+	if len(list) != 1 {
+		return "", false
+	}
+
+	return p.mounted(&list[0])
+}
+
+func (p btrfsPool) mounted(fs *Btrfs) (string, bool) {
+	for _, device := range fs.Devices {
+		if target, ok := getMountTarget(device.Path); ok {
+			return target, true
+		}
+	}
+
+	return "", false
+}
+
+func (p btrfsPool) Name() string {
+	return string(p)
+}
+
+func (p btrfsPool) Path() string {
+	return path.Join("/mnt", string(p))
+}
+
+func (p btrfsPool) enableQuota(mnt string) error {
+	_, err := run(context.Background(), "btrfs", "quota", "enable", mnt)
```

Comment:
aren't you supposed to set the value of the quota too ?

Reply:
No, afaik the value of the quota is set when the Limit method is called, which creates an associated qgroup. This method is called when a pool is mounted, and btrfs will find the associated qgroup and apply the limits.

@muhamadazmy can you confirm this?

Reply:
@LeeSmet yes, that' correct. the enable quota method will only enable quota support for this filesystem. otherwise we can't create qgroups. then later, when we are creating a subvolume, we pass the limit.

Reply:
Go it, thx